### PR TITLE
Fix check for surface crossed when calculating distance to boundary

### PR DESCRIPTION
**Pull request by paulromano**
This is a rather subtle bug that only showed up in a model with very complex cell region definitions. When the distance to the nearest boundary is calculated, an argument `surface_crossed` is returned which gives the index of the surface crossed and its sign indicates whether the particle is going to pass into the positive or negative half-space of that surface. The way the sign is determined is by looking at the region specification of the current cell. If the region specification includes -3 and surface 3 is crossed, it stands to reason that the particle will move into the positive half-space of surface 3, right? Well, this is true for simple cells where the region is defined as only the intersection of half-spaces. In the general case where you can have unions, it is possible that both the negative and positive half-space can appear in a single region specification. As a result, when distance_to_boundary is called, you may get the correct surface_crossed or you may not. For example, here's the cell definition which led me to find this bug:
```xml
 <cell id="21" material="void">
  <region>
    (2 -500 -504) |
    (501 -600 -503) | 
    (600 -502 605 -503) |
    (2 -502 504 -50) | 
    (502 -604 605 -50) | 
    (3 -604 50 -53)
  </region>
</cell>
```

As you can see, both the positive and negative half-spaces of surface 504 (among others) appear in the region specification.

The fix is to explicitly check which half-space the particle is going to move into when it crosses a surface in a complex cell. This is done using the same method in `sense` when the particle is coincident with the surface -- calculate the dot product of the particle direction and the surface normal, and that tells you which half-space the particle is going into.

### CODE REVIEW

File: src/geometry.F90 [hunk 1]
```diff
@@ -598,8 +599,9 @@ subroutine distance_to_boundary(p, dist, surface_crossed, lattice_translation, &
     real(8) :: d_lat              ! distance to lattice boundary
     real(8) :: d_surf             ! distance to surface
     real(8) :: x0,y0,z0           ! coefficients for surface
+    real(8) :: xyz_cross(3)
```
Comment (wbinventor) [inline src/geometry.F90]: if this variable can be explained succinctly in a few-word comment, that would be nice

**Comment by wbinventor** [comment]
This looks fine to me though I think one of our ray tracing gurus like @smharper should check it out first before giving it a pass.

File: src/geometry.F90 (hunk 1 above)
Comment (paulromano) [inline src/geometry.F90]: Good catch- I'll add a comment here.

**Comment by paulromano** [comment]
@smharper let us know if you can take a quick look at this PR. I believe this will be the last thing to go in before the 0.7.1 release.

**Comment by smharper** [comment]
The code looks good, but we should probably modify "test_complex_cells" to check this situation.

**Comment by paulromano** [comment]
Just modified the test to include the situation described. Here's a plot of the new geometry:
1_plot

**Comment by smharper** [comment]
Confirmed that this test fails with the old geometry.F90 and passes with the new one.  Looks like Travis is green so I'm merging now.

**Pull request merged**